instrumentation/opentelemetry-instrumentation-falcon/src/opentelemetry/instrumentation/falcon/__init__.py

```diff
@@ -0,0 +1,192 @@
+from opentelemetry import configuration, context, propagators, trace
+from opentelemetry.instrumentation.falcon.version import __version__
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.utils import http_status_to_canonical_code
+from opentelemetry.trace.status import Status
+from opentelemetry.util import ExcludeList, time_ns
+
+_logger = getLogger(__name__)
+
+_ENVIRON_STARTTIME_KEY = "opentelemetry-falcon.starttime_key"
+_ENVIRON_SPAN_KEY = "opentelemetry-falcon.span_key"
+_ENVIRON_ACTIVATION_KEY = "opentelemetry-falcon.activation_key"
+_ENVIRON_TOKEN = "opentelemetry-falcon.token"
+_ENVIRON_EXC = "opentelemetry-falcon.exc"
+
+
+def get_excluded_urls():
+    urls = configuration.Configuration().FALCON_EXCLUDED_URLS or ""
+    if urls:
+        urls = str.split(urls, ",")
+    return ExcludeList(urls)
+
+
+_excluded_urls = get_excluded_urls()
+
+
+class FalconInstrumentor(BaseInstrumentor):
+    # pylint: disable=protected-access,attribute-defined-outside-init
+    """An instrumentor for falcon.API
+
+    See `BaseInstrumentor`
+    """
+
+    def _instrument(self, **kwargs):
+        self._original_falcon_api = falcon.API
+        falcon.API = _InstrumentedFalconAPI
+
+    def _uninstrument(self, **kwargs):
+        falcon.API = self._original_falcon_api
+
+
+class _InstrumentedFalconAPI(falcon.API):
+    def __init__(self, *args, **kwargs):
+        mw = kwargs.pop("middleware", [])
+        if not isinstance(mw, (list, tuple)):
+            mw = [mw]
+
+        self._tracer = trace.get_tracer(__name__, __version__)
+        mw.insert(0, _TraceMiddleware(self._tracer))
+        kwargs["middleware"] = mw
+        super().__init__(*args, **kwargs)
+
+    def __call__(self, env, start_response):
+        if _excluded_urls.url_disabled(env.get("PATH_INFO", "/")):
+            return super().__call__(env, start_response)
+
+        start_time = time_ns()
+
+        token = context.attach(
+            propagators.extract(otel_wsgi.get_header_from_environ, env)
+        )
+        attributes = otel_wsgi.collect_request_attributes(env)
+        span = self._tracer.start_span(
+            otel_wsgi.get_default_span_name(env),
+            kind=trace.SpanKind.SERVER,
+            attributes=attributes,
+            start_time=start_time,
+        )
+        activation = self._tracer.use_span(span, end_on_exit=True)
+        activation.__enter__()
+        env[_ENVIRON_SPAN_KEY] = span
+        env[_ENVIRON_ACTIVATION_KEY] = activation
+
+        def _start_response(status, response_headers, *args, **kwargs):
+            otel_wsgi.add_response_attributes(span, status, response_headers)
+            activation.__exit__(None, None, None)
+            context.detach(token)
+            return start_response(status, response_headers, *args, **kwargs)
+
+        try:
+            return super().__call__(env, _start_response)
+        except Exception as exc:
+            activation.__exit__(
```

Comment:
I am assuming that line 132 ends up calling `_start_response`. What if an exception is raised during the execution of `start_response` inside `_start_response`? Would that exception be caught here and `activation.__exit__` end up being called twice?

Reply:
Good catch. Fixed to call __exit__ after start_response finishes.